Documentation/learning/why.md

```diff
@@ -71,12 +71,32 @@ If an application reasons primarily about metadata or metadata ordering, such as
 
 For distributed coordination, choosing etcd can help prevent operational headaches and save engineering effort.
 
+### Notes on the usage of lock and lease
+etcd provides [lock APIs][etcd-v3lock] which are based on [the lease mechanism][lease] and [its implementation in etcd][etcdlease]. The basic idea of the lease mechanism is: a server grants a token, which is called lease, to a requesting client. When the server grants a lease, it associates TTL with the lease. When the server detects passing of time longer than TTL, it revokes the lease. During the client holds a non revoked lease, it can claim that it owns an access to a resource associated with the lease. In the case of etcd, the resource is a key in the etcd keyspace. etcd provides lock APIs with this scheme.
+
+The most important aspect of the lease mechanism is that TTL is defined as a physical time interval. Both of the server and client measures passing of time with their own clocks, which are not synchronized each other. It allows a situation that the server revokes the lease but the client still claims it owns the lease because of clock drift.
+
+Then how the lease mechanism guarantees mutual exclusion of the locking mechanism? Actually, the lease mechanism itself doesn't guarantee mutual exclusion. Owning a lease cannot guarantee the owner holds a lock of the resource.
+
+In the case of etcd, mutual exclusion is implemented based on the mechanism of version number validation (it is sometimes called compare and swap in other systems like Consul). In etcd's RPCs like `Put` or `Txn`, we can specify required conditions about revision number and lease ID for the operations. If the conditions are not satisfied, the operation can fail. With this mechanism, etcd provides distributed locking for clients. It means that a client knows that it is acquiring a lock of a key when its requests are completed by etcd cluster successfully.
+
+In distributed locking literature similar designs are described:
+* In [the paper of Chubby][chubby], the concept of *sequencer* is introduced. The authors interpret that sequencer is an almost same to the combination of revision number and lease ID of etcd.
+* In [How to do distributed locking][fencing], Martin Kleppmann introduced the idea of *fencing token*. The authors interpret that fencing token is revision number in the case of etcd. In [Note on fencing and distributed locks][fencing-zk] Flavio Junqueira discussed how the idea of fencing token should be implemented in the case of zookeeper.
+* In [Practical Uses of Synchronized Clocks in Distributed Systems][physicalclock], we can find a description that Thor implements a distributed locking mechanism based on version number validation and lease.
+
+Why do etcd and other systems provide lease if they provide mutual exclusion based on version number validation? Well, leases provide an optimization mechanism for reducing aborted requests.
```

Comment:
This is a key paragraph and neatly sums up the reason for having a lock API.

Reply:
Thanks, I'll emphasize the same message in the first part of the section.

Reply:
A robust lease system can also be used to implement master election without having to reimplement Raft.

That might actually be valuable to be the primary usage example promoted in the documentation.

Reply:
@riking like locking, a naive leader election based on lease mechanism can allow multiple leaders coexist at the same time. I'm not sure it's valuable to add the description here.